### PR TITLE
[v2.7-branch] mgmt: mcumgr: Fix Bluetooth transport issues

### LINKED ISSUE #51846: [v2.7-branch] Bluetooth: Application with buffer that cannot unref it in disconnect handler leads to advertising issues
**Describe the bug**
This issue has been found with a peripheral zephyr device which supports 1 connection, the application works by receiving BT data and then putting it into a fifo to be processed by a dedicated (non-system) workqueue which will send responses to it and unref the connection, an issue arises however in this application if the remote device disconnects prior to the task having finished, the application attempts to restart advertising using the system workqueue from the disconnection handler. The problem is that because the other task contains references to the connection, restarting advertising fails with error -12. Therefore the application is unable to restart advertising and then is in a dead state. Upping the connection count to 2 does work around this issue, but is not an acceptable solution due to the additional RAM usage (which is essentially wasted).
Some operations can also take longer than others, e.g. a image erase command can take upwards of 5 seconds to process.

**To Reproduce**
mcumgr has this issue

**Expected behavior**
To be able to restart advertising, from the system workqueue, triggered in the disconnect handler

**Impact**
Showstopper

**Environment (please complete the following information):**
 - Commit SHA or Version used: This is apparent in an nRF connect SDK sample application (samples/matter/lock) using the softdevice, but should be reproducible with any application that uses bluetooth, I did not attempt seeing if this issue was present in the zephyr controller but would imagine it would be given that it is caused by the limit of connection references/handles being reached

Copy of https://github.com/zephyrproject-rtos/zephyr/issues/51714 but for v2.7-branch.

### PR DESCRIPTION
**Pull request by nordicjm**
This fixes issues with the Bluetooth SMP transport whereby deadlocks could arise from connection references being held in long-lasting mcumgr command processing functions.

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/51714

### CODE REVIEW

**Comment by cfriedt** [comment]
@nordicjm - is this in `main` already?

**Comment by stephanosio** [comment]
The PR to main (https://github.com/zephyrproject-rtos/zephyr/pull/51757) has not been merged yet.

**Review by de-nordic — APPROVED** [review]
OK, although I do not like printk we had to substitute for missing logging....

File: subsys/mgmt/mcumgr/smp_bt.c [hunk 1]
```diff
@@ -51,6 +114,12 @@ static ssize_t smp_bt_chr_write(struct bt_conn *conn,
 {
 	struct smp_bt_user_data *ud;
 	struct net_buf *nb;
+	struct conn_param_data *cpd = conn_param_data_get(conn);
+
+	if (cpd == NULL) {
+		printk("cpd is null");
```
[the remark this answers is not in the document]
Comment (nordicjm) [reply subsys/mgmt/mcumgr/smp_bt.c:120]: I added logging and got some linker error with a dynamic logger error so just used a printk instead, can have another look at it and see if I can resolve it.

[the remark this answers is not in the document]
Comment (de-nordic) [reply subsys/mgmt/mcumgr/smp_bt.c:120]: We don't usually port features down so this may not go through.

[the remark this answers is not in the document]
Comment (nordicjm) [reply subsys/mgmt/mcumgr/smp_bt.c:123]: Fixed

**Review by carlescufi — APPROVED** [review]

**Review by de-nordic — CHANGES_REQUESTED** [review]
I have just noticed that this moves MCUmgr from 2.7 branch to master and rolls back some commits. Why?

**Comment by nordicjm** [comment]
> I have just noticed that this moves MCUmgr from 2.7 branch to master and rolls back some commits. Why?

Do you mean https://github.com/zephyrproject-rtos/zephyr/pull/52248 ? This commit does not change mcumgr repo anymore

**Comment by de-nordic** [comment]
> > I have just noticed that this moves MCUmgr from 2.7 branch to master and rolls back some commits. Why?
> 
> Do you mean #52248 ? This commit does not change mcumgr repo anymore

Is this correct PR: https://github.com/zephyrproject-rtos/mcumgr/pull/37?

**Comment by nordicjm** [comment]
> Is this correct PR: https://github.com/zephyrproject-rtos/mcumgr/pull/37?

No, https://github.com/zephyrproject-rtos/mcumgr/pull/47 is, and https://github.com/zephyrproject-rtos/zephyr/pull/52248 is the PR that moves to the new commit. `v2.7-branch` of zephyr currently points to the top commit of `v2.7-branch` in mcumgr: https://github.com/zephyrproject-rtos/mcumgr/commit/9ffebd5e92d9d069667b9af2a3a028f4a033cfd3

**Comment by de-nordic** [comment]
> > Is this correct PR: [zephyrproject-rtos/mcumgr#37](https://github.com/zephyrproject-rtos/mcumgr/pull/37)?
> 
> No, [zephyrproject-rtos/mcumgr#47](https://github.com/zephyrproject-rtos/mcumgr/pull/47) is, and #52248 is the PR that moves to the new commit. `v2.7-branch` of zephyr currently points to the top commit of `v2.7-branch` in mcumgr: [zephyrproject-rtos/mcumgr@9ffebd5](https://github.com/zephyrproject-rtos/mcumgr/commit/9ffebd5e92d9d069667b9af2a3a028f4a033cfd3)

image

**Comment by nordicjm** [comment]
That was with the earlier west.yml change, that change has been removed from here and moved to https://github.com/zephyrproject-rtos/zephyr/pull/52248

**Comment by nordicjm** [comment]
Point noted and fixed in the other PR.

**Review by de-nordic — APPROVED** [review]

**Pull request merged**